File: RockPaperScissors/javascript/fullrockpaperscissors.js
let readline = require('readline');

let rl = readline.createInterface(
    process.stdin, process.stdout);

function main() {
    giveInstructions();
    playGame();
}


function giveInstructions() {
    console.log("Welcome to the Rock Paper Scissors game!");
    console.log("Here are the rules:");
    console.log("The player and the computer will pick rock, paper, or scissors.");
    console.log("Paper beats rock.");
    console.log("Rock beats scissors");
    console.log("Scissors beat paper.");
}


async function playGame() {
    let playAgain = "Y"
    while(playAgain === "Y" || playAgain === "y") {
        let player = await playersTurn();
        let computer = computersTurn();
        determineWinner(computer, player);
        playAgain = await new Promise(resolve => {
            rl.question('Would you like to play again (Y/N)? ', resolve)
        });
    }
    rl.close();
}


async function playersTurn() {
    let isValid = false;
    let choice = 0;

    while (isValid === false) {
        choice = await new Promise(resolve => {
            rl.question('Enter 1 for Rock, 2 for Paper, 3 for Scissors: ', resolve)
        })
        try {
            choice = parseInt(choice);
            if(choice === 1 || choice === 2 || choice === 3) {
                isValid = true;
            } else {
                console.log("Invalid Input");
            }
        } catch(err) {
            console.log("Invalid Input");
        }
    }

    return choice;
}


function computersTurn() {
    return Math.floor(Math.random() * (3 - 1 + 1) + 1);
}


function convertChoiceToString(intChoice) {
    if(intChoice === 1) {
        return "rock";
    } else if(intChoice === 2) {
        return "paper";
    } else if(intChoice === 3) {
        return "scissors";
    }
}


function determineWinner(computerChoice, userChoice) {
    let stringUserChoice = convertChoiceToString(userChoice);
    let stringComputerChoice = convertChoiceToString(computerChoice);

    let resultString = "The player chose " + stringUserChoice + " and the computer chose " + stringComputerChoice + ", ";

    if(stringUserChoice === stringComputerChoice) {
        console.log(resultString + "it is a tie!");
    } else if(stringUserChoice === "rock") {
        if(stringComputerChoice === "paper") {
            console.log(resultString + "the computer wins!");
        } else if(stringComputerChoice === "scissors") {
            console.log(resultString + "you win!");
        }
    } else if(stringUserChoice === "paper") {
        if(stringComputerChoice === "rock") {
            console.log(resultString + "you win!");
        } else if(stringComputerChoice === "scissors") {
            console.log(resultString + "the computer wins!");
        }
    } else if(stringUserChoice === "scissors") {
        if(stringComputerChoice === "paper") {
            console.log(resultString + "you win!");
        } else if(stringComputerChoice === "rock") {
            console.log(resultString + "the computer wins!");
        }
    }
}


main();

File: RockPaperScissors/python/fullrockpaperscissors.py
import random

def main():
    giveInstructions()
    playGame()


def giveInstructions():
    print("Welcome to the Rock Paper Scissors game!")
    print("Here are the rules:")
    print("The player and the computer will pick rock, paper, or scissors.")
    print("Paper beats rock.")
    print("Rock beats scissors")
    print("Scissors beat paper.")


def playGame():
    play_again = "Y"
    while play_again == "Y" or play_again == "y":
        player = playersTurn()
        computer = computersTurn()
        determineWinner(computer, player)
        play_again = input("Would you like to play again (Y/N)? ")


def playersTurn():
    is_valid = False
    choice = 0

    while is_valid == False:
        user_choice = input("Enter 1 for Rock, 2 for Paper, 3 for Scissors: ")

        try:
            choice = int(user_choice)
            if choice == 1 or choice == 2 or choice == 3:
                is_valid = True
            else:
                print("Invalid input")
        except:
            print("Invalid input")

    return choice


def computersTurn():
    return random.randint(1, 3)


def convertChoiceToString(choice):
    if choice == 1:
        return "rock"
    elif choice == 2:
        return "paper"
    elif choice == 3:
        return "scissors"


def determineWinner(computer_choice, user_choice):
    string_user_choice = convertChoiceToString(user_choice)
    string_computer_choice = convertChoiceToString(computer_choice)
    result_string = "The player chose " + string_user_choice + " and the computer chose " + string_computer_choice + ", "

    if string_user_choice == string_computer_choice:
        print(result_string + "it is a tie!")
    elif string_user_choice == "rock":
        if string_computer_choice == "paper":
            print(result_string + "the computer wins!")
        elif string_computer_choice == "scissors":
            print(result_string + "you win!")
    elif string_user_choice == "paper":
        if string_computer_choice == "rock":
            print(result_string + "you win!")
        elif string_computer_choice == "scissors":
            print(result_string + "the computer wins!")
    elif string_user_choice == "scissors":
        if string_computer_choice == "paper":
            print(result_string + "you win!")
        elif string_computer_choice == "rock":
            print(result_string + "the computer wins!")


main()